File: app/src/main/java/com/pr0gramm/app/ui/views/UserInfoCell.java
package com.pr0gramm.app.ui.views;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.annotation.IdRes;
import android.support.annotation.NonNull;
import android.view.View;
import android.view.ViewParent;
import android.widget.FrameLayout;
import android.widget.TextView;

import com.pr0gramm.app.R;
import com.pr0gramm.app.api.pr0gramm.Api.Info;

import net.danlew.android.joda.DateUtils;

import org.joda.time.Duration;
import org.joda.time.Instant;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 */
@SuppressLint("ViewConstructor")
public class UserInfoCell extends FrameLayout {
    private final UsernameView username;
    private final TextView benis, favorites, comments, tags, uploads;
    private final View messages;
    private final View showComments;
    private final TextView extraInfo;
    private UserActionListener userActionListener;

    public UserInfoCell(Context context, Info userInfo) {
        super(context);
        inflate(context, R.layout.user_info_cell_v2, this);

        username = findView(R.id.username);
        benis = findView(R.id.kpi_benis);
        favorites = findView(R.id.kpi_favorites);
        uploads = findView(R.id.kpi_uploads);
        comments = findView(R.id.kpi_comments);
        tags = findView(R.id.kpi_tags);
        messages = findView(R.id.action_new_message);
        showComments = (View) findView(R.id.kpi_comments).getParent();
        extraInfo = findView(R.id.user_extra_info);

        set(userInfo);
    }

    public void set(Info info) {
        // user info
        Info.User user = info.getUser();
        username.setUsername(user.getName(), user.getMark());
        benis.setText(String.valueOf(user.getScore()));

        // counts
        tags.setText(String.valueOf(info.getTagCount()));
        comments.setText(String.valueOf(info.getCommentCount()));
        uploads.setText(String.valueOf(info.getUploadCount()));

        // open message dialog for user
        messages.setOnClickListener(view -> {
            if (userActionListener != null) {
                userActionListener.onWriteMessageClicked(user.getId(), user.getName());
            }
        });

        ((View) comments.getParent()).setOnClickListener(view -> {
            if (userActionListener != null) {
                userActionListener.onShowCommentsClicked();
            }
        });

        ((View) uploads.getParent()).setOnClickListener(view -> {
            if (userActionListener != null) {
                userActionListener.onShowUploadsClicked(user.getId(), user.getName());
            }
        });

        if (info.likesArePublic()) {
            favorites.setText(String.valueOf(info.getLikeCount()));

            ((View) favorites.getParent()).setOnClickListener(view -> {
                if (userActionListener != null) {
                    userActionListener.onUserFavoritesClicked(user.getName());
                }
            });
        } else {
            // remove the view
            ViewParent parent = favorites.getParent();
            ((View) parent).setVisibility(GONE);
        }

        // info about banned/register date
        if (user.isBanned() != 0) {
            Instant bannedUntil = user.getBannedUntil();
            if (bannedUntil == null) {
                extraInfo.setText(R.string.user_banned_forever);
            } else {
                Duration duration = new Duration(Instant.now(), bannedUntil);
                CharSequence durationStr = DateUtils.formatDuration(getContext(), duration);
                extraInfo.setText(getContext().getString(R.string.user_banned, durationStr));
            }
        } else {
            CharSequence registered = DateUtils.getRelativeTimeSpanString(getContext(),
                    user.getRegistered(), false);

            extraInfo.setText(getContext().getString(R.string.user_registered, registered));
        }
    }

    public void setUserActionListener(UserActionListener userActionListener) {
        this.userActionListener = userActionListener;
    }

    public void setWriteMessageEnabled(boolean enabled) {
        messages.setVisibility(enabled ? VISIBLE : GONE);
    }

    public void setShowCommentsEnabled(boolean enabled) {
        showComments.setVisibility(enabled ? VISIBLE : GONE);
    }

    @NonNull
    @SuppressWarnings("unchecked")
    private <T extends View> T findView(@IdRes int id) {
        View view = checkNotNull(findViewById(id), "view not found");
        return (T) view;
    }

    public interface UserActionListener {
        void onWriteMessageClicked(int userId, String name);

        void onUserFavoritesClicked(String name);

        void onShowCommentsClicked();

        void onShowUploadsClicked(int id, String name);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/views/UserInfoCell.kt
package com.pr0gramm.app.ui.views

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Color
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.github.salomonbrys.kodein.android.appKodein
import com.github.salomonbrys.kodein.instance
import com.jakewharton.rxbinding.view.longClicks
import com.pr0gramm.app.R
import com.pr0gramm.app.UserClasses
import com.pr0gramm.app.api.pr0gramm.Api.Info
import com.pr0gramm.app.services.UriHelper
import com.pr0gramm.app.ui.LoginActivity
import com.pr0gramm.app.util.*
import com.squareup.picasso.Picasso
import kotterknife.bindView
import org.joda.time.Instant
import org.joda.time.Years

/**
 */
@SuppressLint("ViewConstructor")
class UserInfoCell(context: Context, userInfo: Info,
                   private val doIfAuthorizedHelper: LoginActivity.DoIfAuthorizedHelper) : FrameLayout(context) {

    private val username: UsernameView by bindView(R.id.username)
    private val benis: TextView by bindView(R.id.kpi_benis)
    private val favorites: TextView by bindView(R.id.kpi_favorites)
    private val comments: TextView by bindView(R.id.kpi_comments)
    private val tags: TextView by bindView(R.id.kpi_tags)
    private val uploads: TextView by bindView(R.id.kpi_uploads)
    private val extraInfo: TextView by bindView(R.id.user_extra_info)
    private val writeNewMessage: View by bindView(R.id.action_new_message)
    private val writeNewMessageContainer: View by bindView(R.id.action_new_message_container)
    private val actionsContainer: ViewGroup by bindView(R.id.actions_container)
    private val userTypeName: TextView by bindView(R.id.user_type_name)

    private val showCommentsContainer: View

    var userActionListener: UserActionListener? = null

    init {
        View.inflate(context, R.layout.user_info_cell_v2, this)
        showCommentsContainer = find<View>(R.id.kpi_comments).parent as View
        updateUserInfo(userInfo)
    }

    fun updateUserInfo(info: Info) {
        // user info
        val user = info.user
        username.setUsername(user.name, user.mark)
        benis.text = user.score.toString()

        // counts
        tags.text = info.tagCount.toString()
        uploads.text = info.uploadCount.toString()
        comments.text = info.commentCount.toString()

        userTypeName.setTextColor(context.getColorCompat(UserClasses.MarkColors[user.mark]))
        userTypeName.text = context.getString(UserClasses.MarkStrings[user.mark]).toUpperCase()

        // open message dialog for user
        writeNewMessage.setOnClickListener {
            if (userActionListener != null) {
                doIfAuthorizedHelper.run {
                    userActionListener?.onWriteMessageClicked(user.id, user.name)
                }
            }
        }

        (comments.parent as View).setOnClickListener {
            userActionListener?.onShowCommentsClicked()
        }

        (uploads.parent as View).setOnClickListener {
            userActionListener?.onShowUploadsClicked(user.id, user.name)
        }

        if (info.likesArePublic() && info.likeCount > 0) {
            favorites.text = info.likeCount.toString()

            (favorites.parent as View).setOnClickListener {
                userActionListener?.onUserFavoritesClicked(user.name)
            }
        } else {
            // remove the view
            (favorites.parent as View).visibility = View.GONE
        }

        // add badge for "x comments"
        (info.commentCount / 1000).takeIf { it > 0 }?.let {
            appendBadgeView(
                    "comments.png",
                    context.getString(R.string.badge_comments, it.toString()),
                    text = "${it}k",
                    textColor = Color.BLACK)
        }

        // add badge for "x years on pr0gramm"
        Years.yearsBetween(info.user.registered, Instant.now()).years.takeIf { it > 0 }?.let {
            appendBadgeView(
                    "years.png",
                    context.getString(R.string.badge_time, it.toString()),
                    text = it.toString())
        }

        info.badges.forEach { badge ->
            appendBadgeView(badge.image, badge.description ?: "")
        }

        // info about banned/register date
        if (user.isBanned != 0) {
            val bannedUntil = user.bannedUntil
            if (bannedUntil == null) {
                extraInfo.setText(R.string.user_banned_forever)
            } else {
                val durationStr = formatTimeTo(context, bannedUntil, TimeMode.SINCE)
                extraInfo.text = context.getString(R.string.user_banned, durationStr)
            }
        } else {
            val relativeStr = formatTimeTo(context, user.registered, TimeMode.SINCE)
            extraInfo.text = context.getString(R.string.user_registered, relativeStr)
        }
    }

    /**
     * Deflates and composes a badge view and adds that view to the
     * actionsContainer.
     */
    private fun appendBadgeView(image: String, description: String,
                                text: String? = null,
                                textColor: Int? = null) {

        val view = layoutInflater.inflate(R.layout.badge, actionsContainer, false)

        view.longClicks().subscribe {
            Toast.makeText(context, description, Toast.LENGTH_SHORT).show()
        }

        val textView = view.find<TextView>(R.id.text)
        if (text == null) {
            textView.removeFromParent()
        } else {
            textView.text = text
            textColor?.let { textView.setTextColor(it) }
        }

        val imageView = view.find<ImageView>(R.id.image)
        if (!isInEditMode) {
            val picasso = appKodein().instance<Picasso>()

            val localImageId = knownImages.get(image)
            if (localImageId != null) {
                imageView.setImageResource(localImageId)
            } else {
                val url = UriHelper.of(context).badgeImageUrl(image)
                picasso.load(url).into(imageView)
            }
        }

        // add the view to parent
        actionsContainer.addView(view)
    }

    var showWriteMessage: Boolean
        get() = writeNewMessage.visibility == View.VISIBLE
        set(enabled) {
            writeNewMessage.visible = enabled
            writeNewMessageContainer.visible = enabled
        }

    var showComments: Boolean
        get() = showCommentsContainer.visibility == View.VISIBLE
        set(enabled) {
            showCommentsContainer.visible = enabled
        }

    interface UserActionListener {
        fun onWriteMessageClicked(userId: Int, name: String)

        fun onUserFavoritesClicked(name: String)

        fun onShowCommentsClicked()

        fun onShowUploadsClicked(id: Int, name: String)
    }

    companion object {
        private val knownImages = mapOf(
                "years.png" to R.drawable.badge_years,
                "comments.png" to R.drawable.badge_comments,
                "social-share.png" to R.drawable.badge_social,
                "secret-santa-2014.png" to R.drawable.badge_secret_santa,
                "benitrat0r-lose.png" to R.drawable.badge_benitrator_lose,
                "benitrat0r-win.png" to R.drawable.badge_benitrator_win,
                "contract.png" to R.drawable.badge_contract)

    }
}
